File: elixir/calculator.ex
defmodule Calculator do
  # Function to add two numbers
  def add(a, b), do: a + b

  # Function to subtract two numbers
  def subtract(a, b), do: a - b

  # Function to multiply two numbers
  def multiply(a, b), do: a * b

  # Function to divide two numbers
  def divide(a, b) when b != 0, do: a / b
  def divide(_, 0), do: {:error, "Division by zero"}

  # Main function to handle user input and calculations
  def main_loop() do
    IO.puts("Welcome to the Elixir Calculator!")
    IO.puts("Enter an expression (e.g., 2 + 3), or 'quit' to exit:")

    loop()
  end

  defp loop() do
    input = IO.gets(">")
             |> String.trim()

    case input do
      "quit" ->
        IO.puts("Goodbye!")

      _ ->
        case parse_expression(input) do
          {num1, op, num2} ->
            result =
              case op do
                "+" -> add(num1, num2)
                "-" -> subtract(num1, num2)
                "*" -> multiply(num1, num2)
                "/" -> divide(num1, num2)
              end

            IO.puts("Result: #{result}")
          :error ->
            IO.puts("Invalid expression")
        end

        loop()
    end
  end

  defp parse_expression(input) do
    [num1_str, op_str, num2_str] = String.split(input, " ")

    num1 = String.to_float(num1_str)
    num2 = String.to_float(num2_str)

    case {num1, op_str, num2} do
      {n1, op, n2} when is_float(n1) and is_float(n2) -> {n1, op, n2}
      _ -> :error
    end
  end
end

# Start the calculator
Calculator.main_loop()

File: c++/calculator.cpp
#include <iostream>
#include <string>
#include <sstream>

using namespace std;

// Function declarations
double add(double a, double b);
double subtract(double a, double b);
double multiply(double a, double b);
double divide(double a, double b);

int main() {
    cout << "Welcome to the C++ Calculator!" << endl;
    cout << "Enter an expression (e.g., 2 + 3), or 'quit' to exit: " << endl;

    string input;
    while (true) {
        cout << "> ";
        getline(cin, input);

        if (input == "quit") {
            cout << "Goodbye!" << endl;
            break;
        }

        stringstream ss(input);
        double num1, num2;
        char op;
        ss >> num1 >> op >> num2;

        double result;
        switch (op) {
            case '+':
                result = add(num1, num2);
                break;
            case '-':
                result = subtract(num1, num2);
                break;
            case '*':
                result = multiply(num1, num2);
                break;
            case '/':
                result = divide(num1, num2);
                break;
            default:
                cout << "Invalid expression" << endl;
                continue; // Continue to the next iteration of the loop
        }

        cout << "Result: " << result << endl;
    }

    return 0;
}

// Function definitions
double add(double a, double b) {
    return a + b;
}

double subtract(double a, double b) {
    return a - b;
}

double multiply(double a, double b) {
    return a * b;
}

double divide(double a, double b) {
    if (b == 0) {
        cout << "Error: Division by zero" << endl;
        return 0;
    }
    return a / b;
}
